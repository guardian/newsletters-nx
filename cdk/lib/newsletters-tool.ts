import { GuNodeApp } from '@guardian/cdk';
import { AccessScope } from '@guardian/cdk/lib/constants';
import {
	GuDistributionBucketParameter,
	GuStack,
	GuStringParameter,
	type GuStackProps,
} from '@guardian/cdk/lib/constructs/core';
import { GuCname } from '@guardian/cdk/lib/constructs/dns';
import { GuHttpsEgressSecurityGroup } from '@guardian/cdk/lib/constructs/ec2';
import { type App, Duration, SecretValue } from 'aws-cdk-lib';
import { InstanceClass, InstanceSize, InstanceType } from 'aws-cdk-lib/aws-ec2';
import {
	ApplicationListenerRule,
	ListenerAction,
	ListenerCondition,
	UnauthenticatedAction,
} from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { GuPolicy } from '@guardian/cdk/lib/constructs/iam';
import { GuS3Bucket } from '@guardian/cdk/lib/constructs/s3';

export interface NewslettersToolProps extends GuStackProps {
	domainNameTool: string;
	domainNameApi: string;
}

export class NewslettersTool extends GuStack {
	constructor(scope: App, id: string, props: NewslettersToolProps) {
		super(scope, id, props);

		this.setUpNodeEc2(props);
	}

	/**
	 * Generates user data for startup of EC2 instance
	 * User data is a set of instructions to supply to the instance at launch
	 * @see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/user-data.html
	 */
	private getUserData = (
		app: string,
		bucketName: string,
		readOnly: boolean,
	) => {
		// Fetches distribution S3 bucket name from account
		const distributionBucketParameter =
			GuDistributionBucketParameter.getInstance(this);

		return [
			'#!/bin/bash', // "Shebang" to instruct the program loader to run this as a bash script
			'set -e', // Exits immediately if something returns a non-zero status (errors)
			'set +x', // Prevents shell from printing statements before execution
			`aws s3 cp s3://${distributionBucketParameter.valueAsString}/${this.stack}/${this.stage}/${app}/${app}.zip /tmp`, // copies zipped file from s3
			`mkdir -p /opt/${app}`, // make more permanent directory for app to be unzipped into
			`unzip /tmp/${app}.zip -d /opt/${app}`, // unzip the downloaded zip from /tmp into directory in /opt instead
			`chown -R ubuntu /opt/${app}`, // change ownership of the copied files to ubuntu user
			`# write out systemd file
cat >/etc/systemd/system/newsletters-api.service <<EOL
[Unit]
Description=${app}
After=network.target
[Service]
WorkingDirectory=/opt/${app}
Type=simple
User=ubuntu
ExecStart=/usr/bin/node /opt/${app}/dist/apps/newsletters-api/index.cjs
Restart=on-failure
Environment=STAGE=${this.stage}
Environment=NEWSLETTERS_API_READ=${readOnly ? 'true' : 'false'}
Environment=NEWSLETTERS_UI_SERVE=${readOnly ? 'false' : 'true'}
Environment=NEWSLETTER_BUCKET_NAME=${bucketName}
Environment=USE_IN_MEMORY_STORAGE=false
[Install]
WantedBy=multi-user.target
EOL`,
			`systemctl enable newsletters-api`, // enable the service
			`systemctl start newsletters-api`, // start the service
		].join('\n');
	};

	private setUpNodeEc2 = (props: NewslettersToolProps) => {
		const { domainNameTool, domainNameApi } = props;
		const toolAppName = 'newsletters-tool';
		const apiAppName = 'newsletters-api';

		// To avoid exposing the bucket name publicly, fetches the bucket name from SSM (parameter store).
		const bucketSSMParameterName = `/${this.stage}/${this.stack}/${apiAppName}/s3BucketName`;
		const bucketName = StringParameter.valueForStringParameter(
			this,
			bucketSSMParameterName,
		);
		const dataStorageBucket = new GuS3Bucket(this, 'DataBucket', {
			bucketName,
			app: toolAppName,
			versioned: true,
		});

		const s3AccessPolicy = new GuPolicy(this, `s3-access-policy`, {
			policyName: 'readWriteAccessToDataBucket',
			statements: [
				new PolicyStatement({
					sid: 'writeToDataStorageBucketPolicy',
					effect: Effect.ALLOW,
					actions: [
						's3:PutObject',
						's3:GetObject',
						's3:GetObjectVersion',
						's3:DeleteObject',
						's3:ListBucket',
						's3:DeleteObject',
						's3:DeleteObjectVersion',
					],
					resources: [
						`${dataStorageBucket.bucketArn}/*`,
						`${dataStorageBucket.bucketArn}`,
					],
				}),
			],
		});

		/** Sets up Node app to be run in EC2 */
		const ec2AppTool = new GuNodeApp(this, {
			access: { scope: AccessScope.PUBLIC },
			certificateProps: { domainName: domainNameTool },
			monitoringConfiguration: { noMonitoring: true },
			instanceType: InstanceType.of(InstanceClass.T4G, InstanceSize.SMALL),
			// Minimum of 1 EC2 instance running at a time. If one fails, scales up to 2 before dropping back to 1 again
			scaling: { minimumInstances: 1, maximumInstances: 2 },
			// Instructions to set up the environment in the instance
			userData: this.getUserData(toolAppName, bucketName, false),
			roleConfiguration: {
				additionalPolicies: [s3AccessPolicy],
			},
			app: toolAppName,
			accessLogging: {
				enabled: true,
				prefix: `ELBLogs/${this.stack}/${toolAppName}/${this.stage}`,
			},
			applicationLogging: {
				enabled: true,
				systemdUnitName: 'newsletters-api',
			},
		});

		const ec2AppApi = new GuNodeApp(this, {
			access: { scope: AccessScope.PUBLIC },
			certificateProps: { domainName: domainNameApi },
			monitoringConfiguration: { noMonitoring: true },
			instanceType: InstanceType.of(InstanceClass.T4G, InstanceSize.SMALL),
			scaling: { minimumInstances: 1, maximumInstances: 2 },
			userData: this.getUserData(apiAppName, bucketName, true),
			roleConfiguration: {
				additionalPolicies: [s3AccessPolicy],
			},
			app: apiAppName,
			applicationLogging: {
				enabled: true,
				systemdUnitName: 'newsletters-api',
			},
		});

		const readOnlyEndpointApiKeyParam = `/${this.stage}/${this.stack}/${apiAppName}/readOnlyEndpointApiKey`;
		const readOnlyEndpointApiKey = StringParameter.valueForStringParameter(
			this,
			readOnlyEndpointApiKeyParam,
		);

		new ApplicationListenerRule(this, 'ReadOnlyApiHeaderRule', {
			listener: ec2AppApi.listener,
			priority: 1,
			conditions: [
				ListenerCondition.httpHeader('X-Gu-API-Key', [readOnlyEndpointApiKey]),
			],
			targetGroups: [ec2AppApi.targetGroup],
		});
		new ApplicationListenerRule(this, 'BlockRequests', {
			listener: ec2AppApi.listener,
			priority: 2,
			conditions: [ListenerCondition.pathPatterns(['*'])],
			action: ListenerAction.fixedResponse(403, {
				contentType: 'application/json',
				messageBody:
					'{"error": "You are not authorised to access this endpoint"}',
			}),
		});

		/** Security group to allow load balancer to egress to 443 for OIDC flow using Google auth */
		const lbEgressSecurityGroup = new GuHttpsEgressSecurityGroup(
			this,
			'IdP Access',
			{ app: toolAppName, vpc: ec2AppTool.vpc },
		);

		/** Add security group to EC2 load balancer */
		ec2AppTool.loadBalancer.addSecurityGroup(lbEgressSecurityGroup);

		/** Fetch Google clientId param from SSM */
		const clientId = new GuStringParameter(this, 'Google Client ID', {
			description: 'Google OAuth client ID',
			default: `/${this.stage}/${this.stack}/${toolAppName}/googleClientId`,
			fromSSM: true,
		});

		/** Add Google authentication layer to the EC2 load balancer */
		ec2AppTool.listener.addAction('Google Auth', {
			action: ListenerAction.authenticateOidc({
				authorizationEndpoint: 'https://accounts.google.com/o/oauth2/v2/auth',
				issuer: 'https://accounts.google.com',
				scope: 'openid%20profile%20email',
				authenticationRequestExtraParams: { hd: 'guardian.co.uk' },
				onUnauthenticatedRequest: UnauthenticatedAction.AUTHENTICATE,
				tokenEndpoint: 'https://oauth2.googleapis.com/token',
				userInfoEndpoint: 'https://openidconnect.googleapis.com/v1/userinfo',
				clientId: clientId.valueAsString,
				clientSecret: SecretValue.secretsManager(
					`/${this.stage}/deploy/newsletters/clientSecret`,
				),
				next: ListenerAction.forward([ec2AppTool.targetGroup]),
			}),
		});

		/**
		 * Sets up CNAME record for domainName specified
		 * @see https://en.wikipedia.org/wiki/CNAME_record
		 */
		new GuCname(this, 'CNAME', {
			app: toolAppName,
			domainName: domainNameTool,
			ttl: Duration.hours(1),
			resourceRecord: ec2AppTool.loadBalancer.loadBalancerDnsName,
		});
		new GuCname(this, 'NewslettersAPICname', {
			app: apiAppName,
			domainName: domainNameApi,
			ttl: Duration.hours(1),
			resourceRecord: ec2AppApi.loadBalancer.loadBalancerDnsName,
		});
	};
}
