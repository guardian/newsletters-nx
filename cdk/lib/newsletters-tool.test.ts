import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { sharedProps } from '../bin/cdk';
import { NewslettersTool } from './newsletters-tool';

describe('The Newsletters stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new NewslettersTool(app, 'NewslettersTool-TEST', {
			...sharedProps,
			stage: 'TEST',
			app: 'newsletters-tool',
			domainName: 'newsletters-tool.test.dev-gutools.co.uk',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
