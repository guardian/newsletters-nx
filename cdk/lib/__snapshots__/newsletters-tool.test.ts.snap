// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Newsletters stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuVpcParameter",
      "GuSubnetListParameter",
      "GuSubnetListParameter",
      "GuNodeApp",
      "GuCertificate",
      "GuInstanceRole",
      "GuDescribeEC2Policy",
      "GuLoggingStreamNameParameter",
      "GuLogShippingPolicy",
      "GuGetDistributablePolicy",
      "GuParameterStoreReadPolicy",
      "GuAmiParameter",
      "GuHttpsEgressSecurityGroup",
      "GuAutoScalingGroup",
      "GuWazuhAccess",
      "GuApplicationLoadBalancer",
      "GuApplicationTargetGroup",
      "GuHttpsApplicationListener",
      "GuHttpsEgressSecurityGroup",
      "GuStringParameter",
      "GuCname",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "LoadBalancerNewsletterstoolDnsName": {
      "Description": "DNS entry for LoadBalancerNewsletterstool",
      "Value": {
        "Fn::GetAtt": [
          "LoadBalancerNewsletterstoolE9AB350F",
          "DNSName",
        ],
      },
    },
  },
  "Parameters": {
    "AMINewsletterstool": {
      "Description": "Amazon Machine Image ID for the app newsletters-tool. Use this in conjunction with AMIgo to keep AMIs up to date.",
      "Type": "AWS::EC2::Image::Id",
    },
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "GoogleClientID": {
      "Default": "/TEST/newsletters/newsletters-tool/googleClientId",
      "Description": "Google OAuth client ID",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "LoggingStreamName": {
      "Default": "/account/services/logging.stream.name",
      "Description": "SSM parameter containing the Name (not ARN) on the kinesis stream",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "VpcId": {
      "Default": "/account/vpc/primary/id",
      "Description": "Virtual Private Cloud to run EC2 instances within. Should NOT be the account default VPC.",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::VPC::Id>",
    },
    "newsletterstoolPrivateSubnets": {
      "Default": "/account/vpc/primary/subnets/private",
      "Description": "A list of private subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
    "newsletterstoolPublicSubnets": {
      "Default": "/account/vpc/primary/subnets/public",
      "Description": "A list of public subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
  },
  "Resources": {
    "AutoScalingGroupNewsletterstoolASGDA0D2177": {
      "Properties": {
        "HealthCheckGracePeriod": 120,
        "HealthCheckType": "ELB",
        "LaunchConfigurationName": {
          "Ref": "AutoScalingGroupNewsletterstoolLaunchConfigBEDFBC1D",
        },
        "MaxSize": "2",
        "MinSize": "1",
        "Tags": [
          {
            "Key": "App",
            "PropagateAtLaunch": true,
            "Value": "newsletters-tool",
          },
          {
            "Key": "gu:cdk:version",
            "PropagateAtLaunch": true,
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "PropagateAtLaunch": true,
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "LogKinesisStreamName",
            "PropagateAtLaunch": true,
            "Value": {
              "Ref": "LoggingStreamName",
            },
          },
          {
            "Key": "Name",
            "PropagateAtLaunch": true,
            "Value": "NewslettersTool-TEST/AutoScalingGroupNewsletterstool",
          },
          {
            "Key": "Stack",
            "PropagateAtLaunch": true,
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "PropagateAtLaunch": true,
            "Value": "TEST",
          },
        ],
        "TargetGroupARNs": [
          {
            "Ref": "TargetGroupNewsletterstool43BAC68D",
          },
        ],
        "VPCZoneIdentifier": {
          "Ref": "newsletterstoolPrivateSubnets",
        },
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
    },
    "AutoScalingGroupNewsletterstoolInstanceProfile8E2BABCD": {
      "Properties": {
        "Roles": [
          {
            "Ref": "InstanceRoleNewsletterstoolDD0F3665",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "AutoScalingGroupNewsletterstoolLaunchConfigBEDFBC1D": {
      "DependsOn": [
        "InstanceRoleNewsletterstoolDD0F3665",
      ],
      "Properties": {
        "IamInstanceProfile": {
          "Ref": "AutoScalingGroupNewsletterstoolInstanceProfile8E2BABCD",
        },
        "ImageId": {
          "Ref": "AMINewsletterstool",
        },
        "InstanceType": "t4g.small",
        "MetadataOptions": {
          "HttpTokens": "required",
        },
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "GuHttpsEgressSecurityGroupNewsletterstoolA988BFEB",
              "GroupId",
            ],
          },
          {
            "Fn::GetAtt": [
              "WazuhSecurityGroup",
              "GroupId",
            ],
          },
        ],
        "UserData": {
          "Fn::Base64": {
            "Fn::Join": [
              "",
              [
                "#!/bin/bash
set -e
set +x
aws s3 cp s3://",
                {
                  "Ref": "DistributionBucketName",
                },
                "/newsletters/TEST/newsletters-tool/newsletters-tool.zip /tmp
mkdir -p /opt/newsletters-tool
unzip /tmp/newsletters-tool.zip -d /opt/newsletters-tool
chown -R ubuntu /opt/newsletters-tool
su ubuntu -c '/usr/local/node/pm2 start --name newsletters-tool /opt/newsletters-tool/dist/apps/newsletters-api/main.cjs'",
              ],
            ],
          },
        },
      },
      "Type": "AWS::AutoScaling::LaunchConfiguration",
    },
    "CNAME": {
      "Properties": {
        "Name": "newsletters-tool.test.dev-gutools.co.uk",
        "RecordType": "CNAME",
        "ResourceRecords": [
          {
            "Fn::GetAtt": [
              "LoadBalancerNewsletterstoolE9AB350F",
              "DNSName",
            ],
          },
        ],
        "Stage": "TEST",
        "TTL": 3600,
      },
      "Type": "Guardian::DNS::RecordSet",
    },
    "CertificateNewsletterstoolAB12580B": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "DomainName": "newsletters-tool.test.dev-gutools.co.uk",
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters-tool",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Name",
            "Value": "NewslettersTool-TEST/CertificateNewsletterstool",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
      "UpdateReplacePolicy": "Retain",
    },
    "DescribeEC2PolicyFF5F9295": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "autoscaling:DescribeAutoScalingInstances",
                "autoscaling:DescribeAutoScalingGroups",
                "ec2:DescribeTags",
                "ec2:DescribeInstances",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "describe-ec2-policy",
        "Roles": [
          {
            "Ref": "InstanceRoleNewsletterstoolDD0F3665",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetDistributablePolicyNewsletterstool5ADBF705": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:s3:::",
                    {
                      "Ref": "DistributionBucketName",
                    },
                    "/newsletters/TEST/newsletters-tool/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetDistributablePolicyNewsletterstool5ADBF705",
        "Roles": [
          {
            "Ref": "InstanceRoleNewsletterstoolDD0F3665",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GuHttpsEgressSecurityGroupNewsletterstoolA988BFEB": {
      "Properties": {
        "GroupDescription": "Allow all outbound HTTPS traffic",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound HTTPS traffic",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters-tool",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "GuHttpsEgressSecurityGroupNewsletterstoolfromNewslettersToolTESTIdPAccessNewsletterstoolD34AC715300055F7D2AB": {
      "Properties": {
        "Description": "Load balancer to target",
        "FromPort": 3000,
        "GroupId": {
          "Fn::GetAtt": [
            "GuHttpsEgressSecurityGroupNewsletterstoolA988BFEB",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "IdPAccessNewsletterstoolCB424E42",
            "GroupId",
          ],
        },
        "ToPort": 3000,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "GuHttpsEgressSecurityGroupNewsletterstoolfromNewslettersToolTESTLoadBalancerNewsletterstoolSecurityGroup304BBB80300072FF156C": {
      "Properties": {
        "Description": "Load balancer to target",
        "FromPort": 3000,
        "GroupId": {
          "Fn::GetAtt": [
            "GuHttpsEgressSecurityGroupNewsletterstoolA988BFEB",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "LoadBalancerNewsletterstoolSecurityGroup5FD4AD1C",
            "GroupId",
          ],
        },
        "ToPort": 3000,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "GuLogShippingPolicy981BFE5A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "kinesis:Describe*",
                "kinesis:Put*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:kinesis:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":stream/",
                    {
                      "Ref": "LoggingStreamName",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GuLogShippingPolicy981BFE5A",
        "Roles": [
          {
            "Ref": "InstanceRoleNewsletterstoolDD0F3665",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "IdPAccessNewsletterstoolCB424E42": {
      "Properties": {
        "GroupDescription": "Allow all outbound HTTPS traffic",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound HTTPS traffic",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters-tool",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "IdPAccessNewsletterstooltoNewslettersToolTESTGuHttpsEgressSecurityGroupNewsletterstool576F819230004F10AA16": {
      "Properties": {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": {
          "Fn::GetAtt": [
            "GuHttpsEgressSecurityGroupNewsletterstoolA988BFEB",
            "GroupId",
          ],
        },
        "FromPort": 3000,
        "GroupId": {
          "Fn::GetAtt": [
            "IdPAccessNewsletterstoolCB424E42",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 3000,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "InstanceRoleNewsletterstoolDD0F3665": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMManagedInstanceCore",
              ],
            ],
          },
        ],
        "Path": "/",
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters-tool",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ListenerNewsletterstoolFF7146AA": {
      "Properties": {
        "Certificates": [
          {
            "CertificateArn": {
              "Ref": "CertificateNewsletterstoolAB12580B",
            },
          },
        ],
        "DefaultActions": [
          {
            "AuthenticateOidcConfig": {
              "AuthenticationRequestExtraParams": {
                "hd": "guardian.co.uk",
              },
              "AuthorizationEndpoint": "https://accounts.google.com/o/oauth2/v2/auth",
              "ClientId": {
                "Ref": "GoogleClientID",
              },
              "ClientSecret": "{{resolve:secretsmanager:/TEST/deploy/newsletters/clientSecret:SecretString:::}}",
              "Issuer": "https://accounts.google.com",
              "OnUnauthenticatedRequest": "authenticate",
              "Scope": "openid",
              "TokenEndpoint": "https://oauth2.googleapis.com/token",
              "UserInfoEndpoint": "https://openidconnect.googleapis.com/v1/userinfo",
            },
            "Order": 1,
            "Type": "authenticate-oidc",
          },
          {
            "Order": 2,
            "TargetGroupArn": {
              "Ref": "TargetGroupNewsletterstool43BAC68D",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": {
          "Ref": "LoadBalancerNewsletterstoolE9AB350F",
        },
        "Port": 443,
        "Protocol": "HTTPS",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "LoadBalancerNewsletterstoolE9AB350F": {
      "Properties": {
        "LoadBalancerAttributes": [
          {
            "Key": "deletion_protection.enabled",
            "Value": "true",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "LoadBalancerNewsletterstoolSecurityGroup5FD4AD1C",
              "GroupId",
            ],
          },
          {
            "Fn::GetAtt": [
              "IdPAccessNewsletterstoolCB424E42",
              "GroupId",
            ],
          },
        ],
        "Subnets": {
          "Ref": "newsletterstoolPublicSubnets",
        },
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters-tool",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "LoadBalancerNewsletterstoolSecurityGroup5FD4AD1C": {
      "Properties": {
        "GroupDescription": "Automatically created Security Group for ELB NewslettersToolTESTLoadBalancerNewsletterstool4BEF0D7F",
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 443",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters-tool",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "LoadBalancerNewsletterstoolSecurityGrouptoNewslettersToolTESTGuHttpsEgressSecurityGroupNewsletterstool576F81923000859A4066": {
      "Properties": {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": {
          "Fn::GetAtt": [
            "GuHttpsEgressSecurityGroupNewsletterstoolA988BFEB",
            "GroupId",
          ],
        },
        "FromPort": 3000,
        "GroupId": {
          "Fn::GetAtt": [
            "LoadBalancerNewsletterstoolSecurityGroup5FD4AD1C",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 3000,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "ParameterStoreReadNewsletterstoolC810365C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/newsletters/newsletters-tool",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/newsletters/newsletters-tool/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "parameter-store-read-policy",
        "Roles": [
          {
            "Ref": "InstanceRoleNewsletterstoolDD0F3665",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TargetGroupNewsletterstool43BAC68D": {
      "Properties": {
        "HealthCheckIntervalSeconds": 10,
        "HealthCheckPath": "/healthcheck",
        "HealthCheckProtocol": "HTTP",
        "HealthCheckTimeoutSeconds": 5,
        "HealthyThresholdCount": 5,
        "Port": 3000,
        "Protocol": "HTTP",
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters-tool",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "TargetGroupAttributes": [
          {
            "Key": "deregistration_delay.timeout_seconds",
            "Value": "30",
          },
          {
            "Key": "stickiness.enabled",
            "Value": "false",
          },
        ],
        "TargetType": "instance",
        "UnhealthyThresholdCount": 2,
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "WazuhSecurityGroup": {
      "Properties": {
        "GroupDescription": "Allow outbound traffic from wazuh agent to manager",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Wazuh event logging",
            "FromPort": 1514,
            "IpProtocol": "tcp",
            "ToPort": 1514,
          },
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Wazuh agent registration",
            "FromPort": 1515,
            "IpProtocol": "tcp",
            "ToPort": 1515,
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
  },
}
`;
